File: Samples/Strategy Pattern/Sample_1 - CPP/Enemy.cpp
#include "Enemy.h"
#include <iostream>

void Enemy::SetMoveStrategy(IMove* moveStrategy)
{
	if (this->moveStrategy != nullptr)
		delete this->moveStrategy;

	this->moveStrategy = moveStrategy;
}

void Enemy::Move()
{
	if (moveStrategy == nullptr) {
		cout << "MoveStrategy is null!" << endl;
		return;
	}

	moveStrategy->Move();
}

void Enemy::Move(string type)
{
	// Проблема: много вариантов поведения
	// Решение: поведение вынести в отдельные классы
	if (type == "Walk")
		cout << "Enemy walking!" << endl;

	if (type == "Run")
		cout << "Enemy running!" << endl;

	if (type == "Fly")
		cout << "Enemy flying!" << endl;
}

File: Samples/Strategy Pattern/Sample_1 - CS/Enemy.cs
namespace Sample
{
    internal class Enemy
    {
        private IMove? _moveStrategy = null;

        public void SetMoveStrategy(IMove moveStrategy)
        {
            _moveStrategy = moveStrategy;
        }

        public void Move()
        {
            // _moveStrategy?.Move(); 
            // или 
            if (_moveStrategy != null)
                _moveStrategy.Move();
            else
                Console.WriteLine("_moveStrategy is null!");
        }

        public void Move(string type)
        {
            // Проблема: много вариантов поведения
            // Решение: поведение вынести в отдельные классы

            if (type == "Walk")
                Console.WriteLine("Enemy walking!");

            if (type == "Run")
                Console.WriteLine("Enemy running!");

            if (type == "Fly")
                Console.WriteLine("Enemy flying!");
        }
    }
}
